docs/rfcs/0009-remove-interactors-and-controllers.md

```diff
@@ -0,0 +1,193 @@
+
+### Moving into the future: using lib-state with Compose
+
+Ideally, fragments or top-level views/Composables would register observers of their Store's state and send updates from those observers to their child Composables along with closures containing dispatches to those Stores. We are already close to being familiar with this pattern in some places. The following is a current example that has been edited for brevity and clarity:
+
+```kotlin
+class PocketCategoriesViewHolder(
+    composeView: ComposeView,
+    viewLifecycleOwner: LifecycleOwner,
+    private val interactor: PocketStoriesInteractor,
+) : ComposeViewHolder(composeView, viewLifecycleOwner) {
+
+    @Composable
+    override fun Content() {
+        val categories = components.appStore
+            .observeAsComposableState { state -> state.pocketStoriesCategories }.value
+        val categoriesSelections = components.appStore
+            .observeAsComposableState { state -> state.pocketStoriesCategoriesSelections }.value
+
+        PocketTopics(
+            categoryColors = categoryColors,
+            categories = categories ?: emptyList(),
+            categoriesSelections = categoriesSelections ?: emptyList(),
+            onCategoryClick = interactor::onCategoryClicked,
+        )
+    }
+}
+```
+
+Here, we already see a view registering Store observers appropriately. The only thing that would need to change is the line calling the interactor, `onCategoryClick = interactor::onCategoryClicked,`. Following this line leads down a similar chain of abstraction layers, and eventually leads to `DefaultPocketStoriesController::handleCategoryClicked`. This function contains business logic which would need to be moved to the reducer and telemetry side-effects that would need to be moved to a middleware, but the ultimate intent to to toggle the selected category. We skip many layers of indirection by dispatching that action directly from the view:
+
+```kotlin
+class PocketCategoriesViewHolder(
+    composeView: ComposeView,
+    viewLifecycleOwner: LifecycleOwner,
+) : ComposeViewHolder(composeView, viewLifecycleOwner) {
+
+    @Composable
+    override fun Content() {
+        val categories = components.appStore
+            .observeAsComposableState { state -> state.pocketStoriesCategories }.value
+        val categoriesSelections = components.appStore
+            .observeAsComposableState { state -> state.pocketStoriesCategoriesSelections }.value
+
+        PocketTopics(
+            categoryColors = categoryColors,
+            categories = categories ?: emptyList(),
+            categoriesSelections = categoriesSelections ?: emptyList(),
+            onCategoryClick = { name -> 
+                components.appStore.dispatch(AppAction.TogglePocketStoriesCategory(name)) 
```

Comment:
Here, you have made a change to replace the interactor with the dispatch call. Can you add an example of what the other end of the store looks like that connects to this change?

I'm trying to visualize how would features like bookmarks connect to the storage and the snackbar UI layer, or navigate.

Reply:
I had originally intended for this to not necessarily be a deep dive into Redux patterns since there is so much prior art, but I will add some additional examples for clarity.